shared/cancel/canceller.go

```diff
@@ -2,20 +2,60 @@ package cancel
 
 import (
 	"context"
+	"sync"
+	"time"
 )
 
-// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
-// accessible.
+// Canceller behaves similarly to context.WithCancel(), but provides an embedded Cancel function, which makes it
+// more convenient to store in a long-lived struct, and does not use go routines.
+// It implements the context.Context interface.
 type Canceller struct {
-	context.Context
-	Cancel context.CancelFunc
+	ch   chan struct{}
+	once func()
+}
+
+// Err returns nil if Cancel() has not been called yet.
+// If Cancel() has been called then returns context.Canceled.
+// After Err returns a non-nil error, successive calls to Err return the same error.
+func (c *Canceller) Err() error {
+	select {
+	case <-c.ch:
+		return context.Canceled
+	default:
+		return nil
+	}
+}
+
+// Cancel the Canceller.
+// Can be called multiple times safely.
+func (c *Canceller) Cancel() {
+	c.once()
+}
+
+// Done returns a channel that's closed when the canceller.Cancel() is called.
+// Successive calls to Done return the same value.
+func (c *Canceller) Done() <-chan struct{} {
+	return c.ch
+}
+
+// Value is not implemented.
+func (c *Canceller) Value(key any) any {
+	return nil
+}
+
+// Deadline is not implemented.
+func (c *Canceller) Deadline() (deadline time.Time, ok bool) {
+	return time.Time{}, false
 }
 
 // New returns a new canceller with the parent context.
-func New(ctx context.Context) *Canceller {
-	ctx, cancel := context.WithCancel(ctx)
+func New() *Canceller {
```

Comment:
`New returns a new Canceller` (there is no parent context now).

Reply:
fixed